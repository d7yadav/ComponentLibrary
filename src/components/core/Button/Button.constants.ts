import type { ButtonVariant, ButtonSize, ButtonColor } from './Button.types';

export const BUTTON_VARIANTS: Record<ButtonVariant, ButtonVariant> = {
  primary: 'primary',
  secondary: 'secondary',
  tertiary: 'tertiary',
  quaternary: 'quaternary',
  gradient: 'gradient',
  glass: 'glass',
  outline: 'outline',
  text: 'text',
} as const;

export const BUTTON_SIZES: Record<ButtonSize, ButtonSize> = {
  small: 'small',
  medium: 'medium',
  large: 'large',
} as const;

export const BUTTON_COLORS: Record<ButtonColor, ButtonColor> = {
  primary: 'primary',
  secondary: 'secondary',
  tertiary: 'tertiary',
  quaternary: 'quaternary',
  success: 'success',
  warning: 'warning',
  error: 'error',
  info: 'info',
} as const;

export const BUTTON_SIZE_CONFIGS = {
  small: {
    height: 32,
    padding: '6px 12px',
    fontSize: '0.875rem',
    iconSize: 16,
  },
  medium: {
    height: 40,
    padding: '8px 16px',
    fontSize: '1rem',
    iconSize: 20,
  },
  large: {
    height: 48,
    padding: '12px 24px',
    fontSize: '1.125rem',
    iconSize: 24,
  },
} as const;

export const ANIMATION_DURATIONS = {
  hover: '0.2s',
  press: '0.1s',
  loading: '1.5s',
  ripple: '0.6s',
} as const;

export const ANIMATION_EASINGS = {
  spring: 'cubic-bezier(0.175, 0.885, 0.32, 1.275)',
  bounce: 'cubic-bezier(0.68, -0.55, 0.265, 1.55)',
  elastic: 'cubic-bezier(0.25, 0.46, 0.45, 0.94)',
} as const;

export const ACCESSIBILITY_CONSTANTS = {
  minTouchTarget: 44, // Minimum touch target size in pixels
  focusOutlineWidth: 2,
  focusOutlineOffset: 2,
  loadingAriaLabel: 'Loading',
  disabledAriaLabel: 'Button is disabled',
} as const;

export const GLASS_MORPHISM_CONFIG = {
  backdropFilter: 'blur(10px)',
  backdropFilterFallback: 'blur(8px)',
  borderOpacity: 0.2,
  backgroundOpacity: 0.1,
} as const;