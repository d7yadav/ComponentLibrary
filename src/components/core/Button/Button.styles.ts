import type { ButtonProps as MuiButtonProps, PaletteColor } from '@mui/material';
import { Button as MuiButton } from '@mui/material';
import { styled, keyframes } from '@mui/material/styles';

import {
  BUTTON_SIZE_CONFIGS,
  ANIMATION_DURATIONS,
  ANIMATION_EASINGS,
  ACCESSIBILITY_CONSTANTS,
  GLASS_MORPHISM_CONFIG,
} from './Button.constants';
import type { ButtonStyleProps } from './Button.types';
export const StyledButton: React.ComponentType<MuiButtonProps & ButtonStyleProps> = styled(MuiButton, {
  shouldForwardProp: (prop) => !['customVariant', 'customSize', 'customColor', 'loading', 'hasStartIcon', 'hasEndIcon', 'animate'].includes(prop as string),
})<ButtonStyleProps>(
  ({ theme, customVariant, customSize, customColor, loading, disabled, fullWidth, hasStartIcon, hasEndIcon }) => {
    /**
     * Get the size configuration for the current button size
     * @type {typeof BUTTON_SIZE_CONFIGS[typeof customSize]}
     */
    const sizeConfig = BUTTON_SIZE_CONFIGS[customSize];
    /**
     * Determine if the button uses the 'glass' variant
     * @type {boolean}
     */
    const isGlass = customVariant === 'glass';
    /**
     * Determine if the button uses the 'gradient' variant
     * @type {boolean}
     */
    const isGradient = customVariant === 'gradient';
    /**
     * Determine if the button uses the 'outline' variant
     * @type {boolean}
     */
    const isOutline = customVariant === 'outline';
    /**
     * Determine if the button uses the 'text' variant
     * @type {boolean}
     */
    const isText = customVariant === 'text';

    // Base styles
    const baseStyles = {
      'height': sizeConfig.height,
      'padding': sizeConfig.padding,
      'fontSize': sizeConfig.fontSize,
      'fontWeight': 500,
      'fontFamily': theme.typography.fontFamily,
      'textTransform': 'none' as const,
      'borderRadius': theme.shape.borderRadius,
      'minWidth': fullWidth ? '100%' : 'auto',
      'width': fullWidth ? '100%' : 'auto',
      'position': 'relative' as const,
      'overflow': 'hidden' as const,
      'transition': `all ${ANIMATION_DURATIONS.hover} ${ANIMATION_EASINGS.spring}`,
      'cursor': disabled || loading ? 'not-allowed' : 'pointer',
      'opacity': disabled ? 0.6 : 1,
      // Focus styles for accessibility
      '&:focus-visible': {
        'outline': `${ACCESSIBILITY_CONSTANTS.focusOutlineWidth}px solid ${theme.palette.primary.main}`,
        'outlineOffset': ACCESSIBILITY_CONSTANTS.focusOutlineOffset,
      },
      // Icon spacing
      ...(hasStartIcon && {
        '& .MuiButton-startIcon': {
          'marginRight': theme.spacing(1),
          'marginLeft': 0,
        },
      }),
      ...(hasEndIcon && {
        '& .MuiButton-endIcon': {
          'marginLeft': theme.spacing(1),
          'marginRight': 0,
        },
      }),
      // Loading state
      ...(loading && {
        'pointerEvents': 'none' as const,
        '& .MuiButton-startIcon, & .MuiButton-endIcon': {
          'animation': `${loadingSpinKeyframes} ${ANIMATION_DURATIONS.loading} linear infinite`,
        },
      }),
    };

    // Color-based styles
    /**
     * Returns the color-based style object for the button based on its variant and color
     * @returns {React.CSSProperties | Record<string, unknown>} Style object for the button
     */
    const getColorStyles = (): React.CSSProperties | Record<string, unknown> => {
      // Handle custom colors that don't exist in MUI palette
      const getColorPalette = (): PaletteColor => {
        if (customColor === 'tertiary' || customColor === 'quaternary') {
          return theme.palette.primary; // Fallback to primary for custom colors
        }
        return theme.palette[customColor] || theme.palette.primary;
      };
      const colorPalette = getColorPalette();

      if (isText) {
        return {
          'color': colorPalette.main,
          'backgroundColor': 'transparent',
          'border': 'none',
          '&:hover': {
            'backgroundColor': theme.palette.action.hover,
          },
          '&:active': {
            'backgroundColor': theme.palette.action.selected,
          },
        };
      }

      if (isOutline) {
        return {
          'color': colorPalette.main,
          'backgroundColor': 'transparent',
          'border': `1px solid ${colorPalette.main}`,
          '&:hover': {
            'backgroundColor': colorPalette.main,
            'color': colorPalette.contrastText,
          },
          '&:active': {
            'backgroundColor': colorPalette.dark,
          },
        };
      }

      if (isGlass) {
        return {
          'color': colorPalette.main,
          'backgroundColor': `rgba(${theme.palette.mode === 'dark' ? '255, 255, 255' : '0, 0, 0'}, ${GLASS_MORPHISM_CONFIG.backgroundOpacity})`,
          'backdropFilter': GLASS_MORPHISM_CONFIG.backdropFilter,
          'border': `1px solid rgba(${theme.palette.mode === 'dark' ? '255, 255, 255' : '0, 0, 0'}, ${GLASS_MORPHISM_CONFIG.borderOpacity})`,
          '&:hover': {
            'backgroundColor': `rgba(${theme.palette.mode === 'dark' ? '255, 255, 255' : '0, 0, 0'}, ${GLASS_MORPHISM_CONFIG.backgroundOpacity * 1.5})`,
          },
          '@supports not (backdrop-filter: blur())': {
            'backgroundColor': theme.palette.action.hover,
            'border': `1px solid ${theme.palette.divider}`,
          },
        };
      }

      if (isGradient) {
        // Use enhanced theme gradients if available, fallback to basic gradient
        const enhancedTheme: unknown = theme;
        let gradients: Record<string, string> | undefined;
        if (
          typeof enhancedTheme === 'object' &&
          enhancedTheme !== null &&
          'gradients' in enhancedTheme &&
          typeof (enhancedTheme as { gradients?: unknown }).gradients === 'object' &&
          (enhancedTheme as { gradients?: unknown }).gradients !== null
        ) {
          const themeGradients = (enhancedTheme as { gradients?: { dark?: Record<string, string>; light?: Record<string, string> } }).gradients;
          gradients = theme.palette.mode === 'dark' ? themeGradients?.dark : themeGradients?.light;
        }
        const gradient = gradients?.[customColor] || gradients?.['primary'] || `linear-gradient(45deg, ${colorPalette.main}, ${colorPalette.dark})`;
        return {
          'background': gradient,
          'color': theme.palette.getContrastText(colorPalette.main),
          'border': 'none',
          '&:hover': {
            'filter': 'brightness(1.1)',
            'transform': 'translateY(-1px)',
            'boxShadow': theme.shadows[4],
          },
          '&:active': {
            'transform': 'translateY(0px)',
            'filter': 'brightness(0.95)',
          },
        };
      }

      // Solid variants (primary, secondary, tertiary, quaternary)
      return {
        'color': colorPalette.contrastText,
        'backgroundColor': colorPalette.main,
        'border': 'none',
        '&:hover': {
          'backgroundColor': colorPalette.dark,
          'transform': 'translateY(-1px)',
          'boxShadow': theme.shadows[2],
        },
        '&:active': {
          'backgroundColor': colorPalette.dark,
          'transform': 'translateY(0px)',
          'boxShadow': theme.shadows[1],
        },
      };
    };

    return {
      ...baseStyles,
      ...getColorStyles(),
    };
  }
);

// Loading spinner animation
/**
 * Keyframes for loading spinner animation
 * @author dilip.yadav@shorelineiot.com
 */
export const loadingSpinKeyframes = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

/**
 * StyledButtonWithRipple adds a ripple effect for better interaction feedback
 * Filters out all custom props so only valid props are forwarded to the DOM/MUI Button
 * @author dilip.yadav@shorelineiot.com
 */
export const StyledButtonWithRipple: React.ComponentType<MuiButtonProps & ButtonStyleProps> = styled(StyledButton, {
  shouldForwardProp: (prop) => ![
    'customVariant',
    'customSize',
    'customColor',
    'loading',
    'hasStartIcon',
    'hasEndIcon',
    'animate',
    // Do NOT filter out 'startIcon' and 'endIcon' so they reach the underlying MUI Button
    // Add any other custom props here
  ].includes(prop as string),
})(() => ({
  '&:before': {
    'content': '""',
    'position': 'absolute',
    'top': '50%',
    'left': '50%',
    'width': 0,
    'height': 0,
    'borderRadius': '50%',
    'background': 'rgba(255, 255, 255, 0.5)',
    'transform': 'translate(-50%, -50%)',
    'transition': `width ${ANIMATION_DURATIONS.ripple} ease, height ${ANIMATION_DURATIONS.ripple} ease`,
  },
  '&:active:before': {
    'width': '300px',
    'height': '300px',
  },
}));