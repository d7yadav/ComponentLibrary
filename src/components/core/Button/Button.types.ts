import type { ButtonProps as MuiButtonProps } from '@mui/material/Button';
import type { ReactNode } from 'react';

export type ButtonVariant = 
  | 'primary'
  | 'secondary' 
  | 'tertiary'
  | 'quaternary'
  | 'gradient'
  | 'glass'
  | 'outline'
  | 'text';

export type ButtonSize = 'small' | 'medium' | 'large';

export type ButtonColor = 
  | 'primary'
  | 'secondary'
  | 'tertiary'
  | 'quaternary'
  | 'success'
  | 'warning'
  | 'error'
  | 'info';

export interface ButtonProps extends Omit<MuiButtonProps, 'variant' | 'color' | 'size'> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  color?: ButtonColor;
  startIcon?: ReactNode;
  endIcon?: ReactNode;
  loading?: boolean;
  loadingIcon?: ReactNode;
  fullWidth?: boolean;
  disabled?: boolean;
  className?: string;
  children?: ReactNode;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  type?: 'button' | 'submit' | 'reset';
  'aria-label'?: string;
  'aria-describedby'?: string;
  tabIndex?: number;
}

export interface ButtonStyleProps {
  customVariant: ButtonVariant;
  customSize: ButtonSize;
  customColor: ButtonColor;
  loading: boolean;
  disabled: boolean;
  fullWidth: boolean;
  hasStartIcon: boolean;
  hasEndIcon: boolean;
}