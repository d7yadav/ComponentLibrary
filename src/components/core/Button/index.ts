
export { Button } from './Button';
export type { 
  ButtonProps, 
  ButtonVariant, 
  ButtonSize, 
  ButtonColor,
  ButtonStyleProps 
} from './Button.types';
export {
  BUTTON_VARIANTS,
  BUTTON_SIZES,
  BUTTON_COLORS,
  BUTTON_SIZE_CONFIGS,
  ANIMATION_DURATIONS,
  ANIMATION_EASINGS,
  ACCESSIBILITY_CONSTANTS,
  GLASS_MORPHISM_CONFIG,
} from './Button.constants';