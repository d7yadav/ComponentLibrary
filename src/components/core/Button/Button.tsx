import { CircularProgress } from '@mui/material';
import React, { forwardRef, memo } from 'react';
import { isValidElement } from 'react';

import { 
  BUTTON_VARIANTS, 
  BUTTON_SIZES, 
  BUTTON_COLORS,
  ACCESSIBILITY_CONSTANTS 
} from './Button.constants';
import { StyledButtonWithRipple } from './Button.styles';
import type { ButtonProps } from './Button.types';

/**
 * Enhanced Button component with 8 variants, loading states, and accessibility features
 * 
 * Features:
 * - 8 variants: primary, secondary, tertiary, quaternary, gradient, glass, outline, text
 * - Loading states with spinner
 * - Icon support (start/end)
 * - Accessibility compliance (WCAG 2.1 AA)
 * - Dark theme support
 * - Animation effects
 * - Glass morphism and gradient variants
 */
const ButtonComponent = forwardRef<HTMLButtonElement, ButtonProps>(({
  variant = BUTTON_VARIANTS.primary,
  size = BUTTON_SIZES.medium,
  color = BUTTON_COLORS.primary,
  loading = false,
  loadingIcon,
  startIcon,
  endIcon,
  disabled = false,
  fullWidth = false,
  children,
  className,
  onClick,
  type = 'button',
  'aria-label': ariaLabel,
  'aria-describedby': ariaDescribedBy,
  tabIndex,
  ...other
}, ref) => {
  // Determine if button should be disabled
  const isDisabled = disabled || loading;
  
  // Handle loading icon
  /**
   * Returns the loading icon (spinner) for the button
   * @returns {React.ReactNode} The loading spinner icon
   */
  const getLoadingIcon = (): React.ReactNode => {
    if (loadingIcon) {
      return loadingIcon;
    }
    return (
      <CircularProgress
        data-testid="button"
        size={16}
        color="inherit"
        aria-label={ACCESSIBILITY_CONSTANTS.loadingAriaLabel}
      />
    );
  };
  
  // Handle click events
  /**
   * Handles click events for the button
   * @param event React.MouseEvent<HTMLButtonElement>
   * @returns {void}
   */
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>): void => {
    if (isDisabled) {
      event.preventDefault();
      return;
    }
    onClick?.(event);
  };
  
  // Defensive check: Only use valid React elements for icons
  const safeStartIcon = isValidElement(startIcon) ? startIcon : undefined;
  const safeEndIcon = isValidElement(endIcon) ? endIcon : undefined;
  
  // Determine icons to display
  const displayStartIcon = loading ? getLoadingIcon() : safeStartIcon;
  const displayEndIcon = loading ? null : safeEndIcon;
  
  // Accessibility props
  const accessibilityProps = {
    role: 'button',
    'aria-label': ariaLabel || (loading ? ACCESSIBILITY_CONSTANTS.loadingAriaLabel : undefined),
    'aria-describedby': ariaDescribedBy,
    'aria-disabled': isDisabled,
    'aria-busy': loading,
    ...(tabIndex !== undefined && { tabIndex: isDisabled ? -1 : tabIndex }),
  };

  return (
    <StyledButtonWithRipple
      ref={ref}
      variant="text" // Use MUI variant, custom variant handled by styled component props
      size={size === 'small' ? 'small' : size === 'large' ? 'large' : 'medium'}
      // Use 'primary' for 'tertiary' and 'quaternary' to match MUI palette, else use color directly
      color={color === 'tertiary' || color === 'quaternary' ? 'primary' : color}
      // Custom props for styled component (these override the above)
      customVariant={variant}
      customSize={size}
      customColor={color}
      loading={loading}
      disabled={isDisabled}
      fullWidth={fullWidth}
      hasStartIcon={Boolean(displayStartIcon)}
      hasEndIcon={Boolean(displayEndIcon)}
      {...(className && { className })}
      onClick={handleClick}
      type={type}
      startIcon={displayStartIcon}
      endIcon={displayEndIcon}
      {...accessibilityProps}
      {...other}
    >
      {children}
    </StyledButtonWithRipple>
  );
});

ButtonComponent.displayName = 'Button';

// Export memoized component for performance optimization
export const Button = memo(ButtonComponent);