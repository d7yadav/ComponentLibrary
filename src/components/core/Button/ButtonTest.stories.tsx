import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'Test/ButtonSimple',
  component: Button,
  parameters: {
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Default: Story = {
  args: {
    children: 'Test Button',
  },
};

export const Primary: Story = {
  args: {
    children: 'Primary Button',
    variant: 'primary',
  },
};

export const Loading: Story = {
  args: {
    children: 'Loading Button',
    loading: true,
  },
};