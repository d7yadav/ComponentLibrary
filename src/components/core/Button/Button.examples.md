# ðŸ“– Button Examples

## Basic Usage

```typescript
import { Button } from '@/components/core/Button';

// Basic usage
const Example = () => {
  return (
    <Button>
      Basic Button example
    </Button>
  );
};
```

## Advanced Examples

```typescript
// Advanced configuration
const AdvancedExample = () => {
  const [value, setValue] = useState('');
  
  return (
    <Button
      variant="text"
      disabled={false}
      onChange={(e) => setValue(e.target.value)}
      sx={{
        borderRadius: 2,
        '&:hover': {
          transform: 'translateY(-2px)',
        },
      }}
    >
      Advanced Button
    </Button>
  );
};
```

## Integration Examples

```typescript
// Integration with form libraries
import { useController, Control } from 'react-hook-form';

interface FormIntegrationProps {
  name: string;
  control: Control;
}

const FormIntegration = ({ name, control }: FormIntegrationProps) => {
  const { field, fieldState } = useController({
    name,
    control,
    defaultValue: '',
  });

  return (
    <Button
      {...field}
      error={!!fieldState.error}
      helperText={fieldState.error?.message}
    />
  );
};
```

## Edge Cases

```typescript
// Edge case: Empty states
const EmptyState = () => (
  <Button disabled>
    No content available
  </Button>
);

// Edge case: Error states
const ErrorState = () => (
  <Button 
    error
    helperText="Something went wrong"
  />
);

// Edge case: Loading states
const LoadingState = () => (
  <Button 
    loading
    disabled
  />
);
```

## Performance Optimization Examples

```typescript
// Memoized component for performance
const OptimizedButton = React.memo((props: ButtonProps) => {
  const memoizedValue = useMemo(() => {
    // Expensive calculation
    return heavyComputation(props.value);
  }, [props.value]);

  return (
    <Button 
      {...props}
      value={memoizedValue}
    />
  );
});

// Lazy loading for large lists
const LazyList = () => {
  return (
    <VirtualizedList
      itemCount={1000}
      renderItem={({ index }) => (
        <Button key={index}>
          Item {index}
        </Button>
      )}
    />
  );
};
```

## Accessibility Examples

```typescript
// Full accessibility implementation
const AccessibleExample = () => {
  return (
    <Button
      aria-label="Descriptive label for screen readers"
      aria-describedby="helper-text"
      role="button"
      tabIndex={0}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          handleAction();
        }
      }}
    >
      Accessible Button
    </Button>
  );
};

// High contrast mode support
const HighContrastExample = () => {
  return (
    <Button
      sx={{
        '@media (prefers-contrast: high)': {
          borderWidth: 2,
          borderColor: 'text.primary',
        },
      }}
    >
      High contrast support
    </Button>
  );
};
```

---

*These examples demonstrate real-world usage patterns and best practices for the Button component.*
