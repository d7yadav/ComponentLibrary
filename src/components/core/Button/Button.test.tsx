/**
 * @fileoverview Button Component Tests
 * @author AI Generated Tests - 2025-07-13
 */

import { ThemeProvider , createTheme } from '@mui/material/styles';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Button } from './Button';
import { BUTTON_CONSTANTS } from './Button.constants';
import type { ButtonProps } from './Button.types';

// Add jest-axe matcher
expect.extend(toHaveNoViolations);

// Test theme
const testTheme = createTheme();

// Test wrapper component
const TestWrapper = ({ children }: { children: React.ReactNode }) => (
  <ThemeProvider theme={testTheme}>{children}</ThemeProvider>
);

// Helper function to render component with theme
const renderWithTheme = (ui: React.ReactElement, options = {}) => {
  return render(ui, { wrapper: TestWrapper, ...options });
};

describe('Button Component', () => {
  // Basic rendering tests
  describe('Rendering', () => {
    it('renders without crashing', () => {
      renderWithTheme(<Button>Test content</Button>);
      expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('renders children correctly', () => {
      const testContent = 'Test Button Content';
      renderWithTheme(<Button>{testContent}</Button>);
      expect(screen.getByText(testContent)).toBeInTheDocument();
    });

    it('renders with custom test id', () => {
      const testId = 'custom-button';
      renderWithTheme(
        <Button data-testid={testId}>Test</Button>
      );
      expect(screen.getByTestId(testId)).toBeInTheDocument();
    });

    it('applies custom className', () => {
      const customClass = 'custom-class';
      renderWithTheme(
        <Button className={customClass}>Test</Button>
      );
      expect(screen.getByRole('button')).toHaveClass(customClass);
    });
  });

  // Props testing
  describe('Props', () => {
    describe('variant prop', () => {
      const variants: ButtonProps['variant'][] = [
        'primary',
        'secondary',
        'tertiary',
        'quaternary',
        'gradient',
        'glass',
        'outline',
        'text',
      ];

      variants.forEach((variant) => {
        it(`renders ${variant} variant correctly`, () => {
          renderWithTheme(
            <Button variant={variant} data-testid="variant-test">
              {variant} variant
            </Button>
          );
          const element = screen.getByTestId('variant-test');
          expect(element).toBeInTheDocument();
        });
      });
    });

    describe('size prop', () => {
      const sizes: ButtonProps['size'][] = [
        'small',
        'medium',
        'large',
      ];

      sizes.forEach((size) => {
        it(`renders ${size} size correctly`, () => {
          renderWithTheme(
            <Button size={size} data-testid="size-test">
              {size} size
            </Button>
          );
          const element = screen.getByTestId('size-test');
          expect(element).toBeInTheDocument();
        });
      });
    });


    it('handles loading state correctly', () => {
      renderWithTheme(
        <Button loading data-testid="loading-test">
          Loading Button
        </Button>
      );
      
      const element = screen.getByTestId('loading-test');
      expect(element).toHaveAttribute('aria-busy', 'true');
    });

    it('handles disabled state correctly', () => {
      const handleClick = vi.fn();
      renderWithTheme(
        <Button disabled onClick={handleClick} data-testid="disabled-test">
          Disabled Button
        </Button>
      );
      
      const element = screen.getByTestId('disabled-test');
      expect(element).toHaveAttribute('aria-disabled', 'true');
    });
  });

  // Interaction testing
  describe('Interactions', () => {
    it('handles click events', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Button onClick={handleClick}>
          Clickable Button
        </Button>
      );
      
      await user.click(screen.getByRole('button'));
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('handles keyboard navigation - Enter key', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Button onClick={handleClick}>
          Keyboard Button
        </Button>
      );
      
      const button = screen.getByRole('button');
      button.focus();
      await user.keyboard('{Enter}');
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('handles keyboard navigation - Space key', async () => {
      const user = userEvent.setup();
      const handleClick = vi.fn();
      
      renderWithTheme(
        <Button onClick={handleClick}>
          Keyboard Button
        </Button>
      );
      
      const button = screen.getByRole('button');
      button.focus();
      await user.keyboard(' ');
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

  });


  // Icon testing
  describe('Icons', () => {
    const TestIcon = () => <span data-testid="test-icon">Icon</span>;

    it('renders start icon correctly', () => {
      renderWithTheme(
        <Button startIcon={<TestIcon />}>
          With Start Icon
        </Button>
      );
      
      expect(screen.getByTestId('test-icon')).toBeInTheDocument();
    });

    it('renders end icon correctly', () => {
      renderWithTheme(
        <Button endIcon={<TestIcon />}>
          With End Icon
        </Button>
      );
      
      expect(screen.getByTestId('test-icon')).toBeInTheDocument();
    });
  });

  // Accessibility testing
  describe('Accessibility', () => {
    it('meets WCAG accessibility guidelines', async () => {
      const { container } = renderWithTheme(
        <Button>Accessible Button</Button>
      );
      
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });

    it('supports screen readers', () => {
      const ariaLabel = 'Screen reader label';
      renderWithTheme(
        <Button aria-label={ariaLabel}>
          Button for screen readers
        </Button>
      );
      
      expect(screen.getByLabelText(ariaLabel)).toBeInTheDocument();
    });

    it('supports keyboard navigation', () => {
      renderWithTheme(<Button>Keyboard Button</Button>);
      
      const element = screen.getByRole('button');
      expect(element).toHaveAttribute('tabIndex', '0');
    });

    it('provides proper ARIA attributes', () => {
      renderWithTheme(
        <Button 
          aria-describedby="description"
          aria-controls="controlled-element"
        >
          ARIA Button
        </Button>
      );
      
      const element = screen.getByRole('button');
      expect(element).toHaveAttribute('aria-describedby', 'description');
      expect(element).toHaveAttribute('aria-controls', 'controlled-element');
    });
  });

  // Theme integration testing
  describe('Theme Integration', () => {
    it('respects theme customizations', () => {
      const customTheme = createTheme({
        components: {
          MuiButton: {
            defaultProps: {
              variant: 'primary',
            },
            styleOverrides: {
              root: {
                backgroundColor: 'rgb(255, 0, 0)',
              },
            },
          },
        },
      });

      render(
        <ThemeProvider theme={customTheme}>
          <Button data-testid="themed-button">
            Themed Button
          </Button>
        </ThemeProvider>
      );

      const element = screen.getByTestId('themed-button');
      expect(element).toBeInTheDocument();
    });

    it('supports dark mode', () => {
      const darkTheme = createTheme({
        palette: {
          mode: 'dark',
        },
      });

      render(
        <ThemeProvider theme={darkTheme}>
          <Button>Dark Mode Button</Button>
        </ThemeProvider>
      );

      expect(screen.getByRole('button')).toBeInTheDocument();
    });
  });

  // Performance testing
  describe('Performance', () => {
    it('renders efficiently with many instances', () => {
      const manyComponents = Array.from({ length: 100 }, (_, i) => (
        <Button key={i}>Button {i}</Button>
      ));

      const { container } = renderWithTheme(<div>{manyComponents}</div>);
      expect(container.querySelectorAll('[role="button"]')).toHaveLength(100);
    });

    it('respects reduced motion preferences', () => {
      // Mock prefers-reduced-motion
      Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: vi.fn().mockImplementation((query) => ({
          matches: query === '(prefers-reduced-motion: reduce)',
          media: query,
          onchange: null,
          addListener: vi.fn(),
          removeListener: vi.fn(),
          addEventListener: vi.fn(),
          removeEventListener: vi.fn(),
          dispatchEvent: vi.fn(),
        })),
      });

      renderWithTheme(
        <Button animate="false">
          No Animation Button
        </Button>
      );

      expect(screen.getByRole('button')).toBeInTheDocument();
    });
  });

  // Edge cases and error handling
  describe('Edge Cases', () => {
    it('handles empty children gracefully', () => {
      renderWithTheme(<Button />);
      expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('handles null children gracefully', () => {
      renderWithTheme(<Button>{null}</Button>);
      expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('handles undefined props gracefully', () => {
      renderWithTheme(
        <Button 
          variant={undefined as any} 
          size={undefined as any}
        >
          Undefined Props
        </Button>
      );
      expect(screen.getByRole('button')).toBeInTheDocument();
    });

  });

  // Snapshot testing
  describe('Snapshots', () => {
    it('matches snapshot for default props', () => {
      const { container } = renderWithTheme(
        <Button>Default Button</Button>
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches snapshot for all variants', () => {
      const variants: ButtonProps['variant'][] = [
        'primary',
        'secondary',
        'tertiary',
        'quaternary',
        'gradient',
        'glass',
        'outline',
        'text',
      ];

      variants.forEach((variant) => {
        const { container } = renderWithTheme(
          <Button variant={variant}>{variant} Button</Button>
        );
        expect(container.firstChild).toMatchSnapshot(`variant-${variant}`);
      });
    });
  });
});
