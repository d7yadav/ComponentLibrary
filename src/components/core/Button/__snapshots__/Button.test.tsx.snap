// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-glass 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-gradient 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-outline 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-primary 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-quaternary 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-secondary 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-tertiary 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for all variants > variant-text 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;

exports[`Button Component > Snapshots > matches snapshot for default props 1`] = `
<style>
  
  @keyframes spin {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }

</style>
`;
